### PR TITLE
[Linux] new testcase vtpm_cold_add_remove for linux

### PR DESCRIPTION
**Pull request by ZouYuhua**
Task:
new testcase vtpm_cold_add_remove for linux: GOSV-2531

### CODE REVIEW

File: linux/utils/get_tpm_status.yml [hunk 1]
```diff
@@ -0,0 +1,76 @@
+# Copyright 2025 VMware, Inc.
+# SPDX-License-Identifier: BSD-2-Clause
+---
+# Get TPM device info in Windows guest OS
+# Return:
+#   tpm_present: the vTPM is present or not
```
Comment (Tomorrow9) [inline linux/utils/get_tpm_status.yml]: vTPM -> TPM device

File: linux/utils/get_tpm_status.yml [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  ansible.builtin.debug: var=tpm_dev_result
+
+- name: "Set TPM device present or not"
+  ansible.builtin.set_fact:
+    tpm_present: true
+  when:
+    - tpm_dev_result.failed is defined
+    - not tpm_dev_result.failed
+    - tpm_dev_result.stdout is defined
+    - tpm_dev_result.stdout
```
Comment (Tomorrow9) [inline linux/utils/get_tpm_status.yml]: Use `tpm_dev_result.stdout | length != 0` instead to align with the requirement in Ansible 2.19. And other `.stdout` in this task file.

File: linux/vtpm_cold_add_remove/clear_tpm_test.yml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# Copyright 2025 VMware, Inc.
+# SPDX-License-Identifier: BSD-2-Clause
+---
+# Check OS still work well after clear TPM
+#
+- name: "Clear vTPM in guest OS"
```
Comment (Tomorrow9) [inline linux/vtpm_cold_add_remove/clear_tpm_test.yml]: vTPM -> TPM

File: linux/vtpm_cold_add_remove/clear_tpm_test.yml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  ansible.builtin.shell: tpm2_clear
+  delegate_to: "{{ vm_guest_ip }}"
+
+- name: "Restart guest OS after reset TPM"
```
Comment (Tomorrow9) [inline linux/vtpm_cold_add_remove/clear_tpm_test.yml]: Need to use the task file in utils to restart guest OS since VMware Tools can be not available.

File: linux/vtpm_cold_add_remove/vtpm2_activatecredential_test.yml [hunk 5]
```diff
@@ -0,0 +1,55 @@
+# Copyright 2025 VMware, Inc.
+# SPDX-License-Identifier: BSD-2-Clause
+---
+#
+- name: "Check supporting ActivateCredential or not"
```
Comment (Tomorrow9) [inline linux/vtpm_cold_add_remove/vtpm2_activatecredential_test.yml]: These multiple tasks can be put in a single shell script or in a single task with shell commands to avoid the whole testing disrupted by connection issue.

File: linux/utils/get_tpm_status.yml [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+- name: "Add manafacturer info to dict tpm_info"
+  ansible.builtin.set_fact:
+    tpm_info: "{{ tpm_info | combine({'TPM2_PT_MANUFACTURER': tpm_manufacturer_results.stdout}) }}"
+  when:
+    - tpm_manufacturer_results.failed is defined
+    - not tpm_manufacturer_results.failed
+    - tpm_manufacturer_results.stdout is defined
+    - tpm_manufacturer_results.stdout | length != 0
+
+- name: "Get vTPM family"
```
Comment (Tomorrow9) [inline linux/utils/get_tpm_status.yml]: vTPM -> TPM

File: linux/utils/get_tpm_status.yml [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  ansible.builtin.shell: tpm2_getcap properties-fixed | awk '/TPM2_PT_FAMILY_INDICATOR:/ {getline;getline; gsub(/"/, "", $2); print $2}'
+  delegate_to: "{{ vm_guest_ip }}"
+  register: tpm_family_results
+
+- name: "Add vTPM family info to dict tpm_info"
```
Comment (Tomorrow9) [inline linux/utils/get_tpm_status.yml]: vTPM -> TPM

File: linux/vtpm_cold_add_remove/clear_tpm_test.yml [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+# Check OS still work well after clear TPM
+#
+- name: "Clear TPM in guest OS"
+  ansible.builtin.shell: tpm2_clear
+  delegate_to: "{{ vm_guest_ip }}"
+
+- name: "Reboot guest OS"
+  include_tasks: ../utils/reboot.yml
+
+#- name: "Restart guest OS after reset TPM"
```
Comment (Tomorrow9) [inline linux/vtpm_cold_add_remove/clear_tpm_test.yml]: The commented lines can be removed, and need to add `- name: "Update VM's guest IP after reboot"
  include_tasks: ../../common/update_inventory.yml` after reboot task.

File: linux/vtpm_cold_add_remove/vtpm2_hash_test.yml [hunk 9]
```diff
@@ -0,0 +1,21 @@
+# Copyright 2025 VMware, Inc.
+# SPDX-License-Identifier: BSD-2-Clause
+---
+#
```
Comment (Tomorrow9) [inline linux/vtpm_cold_add_remove/vtpm2_hash_test.yml]: You may add the description of this test purpose instead of the empty line.

File: linux/vtpm_cold_add_remove/vtpm2_pcr_consistency_test.yml [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+  include_tasks: ../../common/esxi_get_cpu_mem_info.yml
+
+- name: "Set fact of the maximum vCPU number of hot add test"
+  ansible.builtin.set_fact:
+    max_num_cpus: "{{ [esxi_logical_cpu_num | int, vm_cpu_hotadd_max | default(16) | int] | min }}"
+
+- name: "Maximum VM vCPU number of CPU hot add test"
+  ansible.builtin.debug: var=max_num_cpus
+
+- name: "Get CPU info"
```
Comment (Tomorrow9) [inline linux/vtpm_cold_add_remove/vtpm2_pcr_consistency_test.yml]: Above tasks are not necessary, since you do not do CPU add test, if you just want to change the VM CPU config, just switch the values, for example, CPU number is the same, change the CPU cores per socket number.

File: linux/vtpm_cold_add_remove/vtpm_install_packages.yml [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+---
+#
+- name: "Install 'tpm2-tools' for {{ guest_os_ansible_distribution }}"
+  include_tasks: ../utils/install_uninstall_package.yml
+  vars:
+    package_list: ["tpm2-tools"]
+    package_state: "present"
+  when: guest_os_ansible_distribution not in ['SLES', 'SLED', 'openSUSE Leap']
+
+# Configure local DVD repositories to install packages for openSUSE, SLES.
```
Comment (Tomorrow9) [inline linux/vtpm_cold_add_remove/vtpm_install_packages.yml]: This is config repo tasks in `install_uninstall_package.yml`

File: linux/utils/get_tpm_status.yml [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+- name: "Set TPM device present or not"
+  ansible.builtin.set_fact:
+    tpm_present: true
+  when:
+    - tpm_dev_result.failed is defined
+    - not tpm_dev_result.failed
+    - tpm_dev_result.stdout is defined
+    - tpm_dev_result.stdout | length != 0
+
+- name: "Get TPM info"
```
Comment (Tomorrow9) [inline linux/utils/get_tpm_status.yml]: The following tasks only be executed when `tpm_present` is true.

File: linux/vtpm_cold_add_remove/clear_tpm_test.yml [hunk 13]
```diff
@@ -0,0 +1,17 @@
+# Copyright 2026 VMware, Inc.
+# SPDX-License-Identifier: BSD-2-Clause
+---
+# Check OS still work well after clear TPM
+#
+- name: "Clear TPM in guest OS"
+  ansible.builtin.shell: tpm2_clear
+  delegate_to: "{{ vm_guest_ip }}"
+
+- name: "Reboot guest OS"
```
Comment (Tomorrow9) [inline linux/vtpm_cold_add_remove/clear_tpm_test.yml]: Can pause several seconds after executing above command and before reboot.

File: linux/vtpm_cold_add_remove/vtpm2_pcr_consistency_test.yml [hunk 14]
```diff
@@ -0,0 +1,63 @@
+# Copyright 2026 VMware, Inc.
+# SPDX-License-Identifier: BSD-2-Clause
+---
+- name: "Set the vtpm_cpus and vtpm_cpu_cores_per_socket facts to their default values"
```
Comment (Tomorrow9) [inline linux/vtpm_cold_add_remove/vtpm2_pcr_consistency_test.yml]: >- "Initialize the changed VM CPU and cores per socket number to the VM default config values"

File: linux/vtpm_cold_add_remove/vtpm2_pcr_consistency_test.yml [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+  ansible.builtin.set_fact:
+    vtpm_cpus: "{{ (guest_config_options.rec_cpu_cores_per_socket | int * guest_config_options.rec_cpu_socket | int) | default(2) }}"
+    vtpm_cpu_cores_per_socket: "{{ guest_config_options.rec_cpu_cores_per_socket | default(1) }}"
+
+- name: "Get CPU info"
+  include_tasks: ../../common/vm_get_cpu_info.yml
+
+- name: "Set the vtpm_num_cpus fact when the current CPU count and cores per socket match the default values"
```
Comment (Tomorrow9) [inline linux/vtpm_cold_add_remove/vtpm2_pcr_consistency_test.yml]: >- "Set changed VM CPU number when the VM current config matches the VM default config values"

File: linux/vtpm_cold_add_remove/vtpm2_pcr_consistency_test.yml [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+  ansible.builtin.set_fact:
+    vtpm_num_cpus: "{{ vtpm_cpus + vtpm_cpu_cores_per_socket }}"
+  when:
+    - vm_cpu_num == vtpm_cpus
+    - vm_cpu_cores_per_socket == vtpm_cpu_cores_per_socket
+
+- name: "Change vCPU and test vTPM.enable.pcr_consistency.ChangeHardware"
```
Comment (Tomorrow9) [inline linux/vtpm_cold_add_remove/vtpm2_pcr_consistency_test.yml]: "vTPM.enable.pcr_consistency.ChangeHardware" should be changed to the description.

File: linux/vtpm_cold_add_remove/vtpm2_pcr_consistency_test.yml [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+      delegate_to: "{{ vm_guest_ip }}"
+      register: pcr_before_result
+
+    - name: "Display current PCR value"
+      ansible.builtin.debug: var=pcr_before_result
+
+    - name: "Shutdown VM"
+      include_tasks: ../utils/shutdown.yml
+
+    - name: "Modify VM settting to add vCPU"
```
Comment (Tomorrow9) [inline linux/vtpm_cold_add_remove/vtpm2_pcr_consistency_test.yml]: "Change VM vCPU number"

File: linux/vtpm_cold_add_remove/vtpm_check_status.yml [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+# SPDX-License-Identifier: BSD-2-Clause
+---
+# Check vTPM device status in guest OS
+- name: "Make sure TPM device is present"
+  ansible.builtin.assert:
+    that:
+      - tpm_present | bool
+    fail_msg: "TPM device status is not present in guest OS {{ guest_os_ansible_distribution }}."
+
+- name: "Get random numbers form TPM"
```
Comment (Tomorrow9) [inline linux/vtpm_cold_add_remove/vtpm_check_status.yml]: > from

File: linux/vtpm_cold_add_remove/vtpm_install_packages.yml [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+      register: tpm_tools_version_result
+
+    - name: "Set fact of tpm2-tools major version"
+      ansible.builtin.set_fact:
+        tpm2_major_version: "{{ (tpm_tools_version_result | regex_search('tpm2-tools-([0-9]+)', '\\1'))[0] }}"
+
+- name: "Print tpm2-tools major version"
+  ansible.builtin.debug: var=tpm2_major_version
+
+- name: "Install package openssl for {{ guest_os_ansible_distribution }}"
```
Comment (Tomorrow9) [inline linux/vtpm_cold_add_remove/vtpm_install_packages.yml]: This install packages can be moved at the beginning as others. And you can defined the packages list for different OS distr and call `install_uninstall_package.yml` once.

File: linux/vtpm_cold_add_remove/vtpm_add_remove_check.yml [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+
+- name: "Check TPM device status in guest OS after cold add"
+  include_tasks: vtpm_check_status.yml
+  when: vtpm_test_operation == 'add'
+
+- name: "Check TPM device is not present after cold remove"
+  ansible.builtin.assert:
+    that:
+      - tpm_present is defined
+      - not tpm_present | bool
```
[the remark this answers is not in the document]
Comment (ZouYuhua) [reply linux/vtpm_cold_add_remove/vtpm_add_remove_check.yml]: If the vTPM device is removed, the device descriptor files /dev/tpm* does not exist. 
If the vTPM device exists, It will has device descriptor files /dev/tpm*
This is the intended behavior of the kernel.

**Review by Tomorrow9 — APPROVED** [review]
LGTM, thanks!
